File: src/main/scala/edu/sjsu/mithai/graphX/GraphProcessor.java
//package edu.sjsu.mithai.graphX;
//
//import org.apache.spark.graphx.Graph;
//
//import java.util.ArrayList;
//
///**
// * Created by Madhura on 9/26/16.
// */
//
//public class GraphProcessor {
//
//    public static void main(String[] args) {
//
//        //incoming arrayliist (call from kaustubh`s func)
//        ArrayList<String> vertex = new ArrayList<>();
//        vertex.add("{\"sensor1\",\"100\"}");
//        vertex.add("{\"sensor2\",\"200\"}");
//        vertex.add("{\"sensor3\",\"300\"}");
//        vertex.add("{\"sensor4\",\"400\"}");
//        vertex.add("{\"sensor5\",\"500\"}");
//        vertex.add("{\"sensor6\",\"600\"}");
//
//        //generated graph
//        Graph<String, Object> graph = new GraphCreator().createGraph(vertex);
//
//        System.out.println("Graph Object ------->>>>>"+graph);
//
//        //TODO: Add graph processing functions
//        System.out.println(graph.aggregateMessages$default$3());
//        System.out.println(graph.vertices().collect());
//
//    }
//
//
//
//}

File: src/main/scala/edu/sjsu/mithai/graphX/GraphProcessor.scala
package edu.sjsu.mithai.graphX

import org.apache.spark.graphx._
import org.apache.spark.graphx.lib.ShortestPaths

object GraphProcessor {

  def min(graph: Graph[(String, Double), PartitionID]): (VertexId, (String, Double)) = {
    return graph.vertices.min()(new Ordering[Tuple2[VertexId, (String, Double)]]() {
      override def compare(x: (VertexId, (String, Double)), y: (VertexId, (String, Double))): Int =
        Ordering[Double].compare(x._2._2, y._2._2)
    })
  }

  def max(graph: Graph[(String, Double), PartitionID]): (VertexId, (String, Double)) = {
    return graph.vertices.max()(new Ordering[Tuple2[VertexId, (String, Double)]]() {
      override def compare(x: (VertexId, (String, Double)), y: (VertexId, (String, Double))): Int =
        Ordering[Double].compare(x._2._2, y._2._2)
    })
  }

  def shortestPath(graph: Graph[(String, Double), PartitionID], entry: String): String = {

    //TODO : List entry points in the graph (Need to add identifiers in graph properties)
    val firstVId = entry.hashCode

    println("Vertex ID --->" + firstVId)
    val vert: Seq[VertexId] = graph.vertices.collect().map(x=>x._1).toSeq
    val result = ShortestPaths.run(graph, vert)
    if (result.vertices != null && result.vertices.count() > 1) {
      val shortPath = result.vertices.filter(
        {
          case (vid, path) => {
            vid == firstVId
          }
        }).first()._2
      //Finds the non zero min value
      var second = shortPath.filterNot(_._2 == 0).min
      println("Shortest Path from vertices " + entry + " is: " + second._1 + " with length: " + second._2)
      return second._1.toString
    }
    return null;
  }

  def average(graph: Graph[(String, Double), PartitionID]): Double = {
    return graph.vertices.map(_._2._2).sum() / graph.vertices.count()
  }

  def graphOperation(f:Graph[(String, Double), PartitionID] => (VertexId, (String, Double)), graph: Graph[(String, Double), PartitionID]): (VertexId, (String, Double)) = {
    if (graph!=null) {
      val result = f(graph);
      return result;
    }
    else return f(graph);
  }

}


